crates/sparrow-merge/src/merge_pipeline.rs

```diff
@@ -0,0 +1,316 @@
+use arrow_schema::DataType;
+use error_stack::{IntoReport, ResultExt};
+use parking_lot::Mutex;
+use sparrow_batch::Batch;
+use sparrow_physical::StepId;
+use sparrow_scheduler::{
+    InputHandles, Partition, Partitioned, Pipeline, PipelineError, Scheduler, TaskRef,
+};
+use std::sync::atomic::{AtomicBool, Ordering};
+use tokio::sync::mpsc::error::TryRecvError;
+
+use crate::merge::HeterogeneousMerge;
+
+/// Runs a merge operation.
+pub struct MergePipeline {
+    /// The state for each partition.
+    partitions: Partitioned<MergePartition>,
+    /// Input handles for the consuming (receiving) computations.
+    consumers: InputHandles,
+    /// The left input.
+    left: Input,
+    /// The right input.
+    right: Input,
+    /// The result type of this merge.
+    result_type: DataType,
+}
+
+impl std::fmt::Debug for MergePipeline {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MergePipeline").finish()
+    }
+}
+
+struct Input {
+    /// The input id for this input
+    #[allow(unused)]
+    id: StepId,
+    /// The data type for this input
+    datatype: DataType,
+    /// Whether this side is closed
+    is_closed: Mutex<bool>,
+}
+
+struct MergePartition {
+    /// Whether this partition is closed.
+    is_closed: AtomicBool,
+    /// Task for this partition.
+    task: TaskRef,
+    /// The senders for each input.
+    ///
+    /// Batches are first placed into the channels, which allows us to use
+    /// [Gatherer] to choose which input needs to be merged next.
+    ///
+    /// This vec should contains one element for each input.
+    txs: Vec<tokio::sync::mpsc::UnboundedSender<Batch>>,
+    // Keeps track of the entities seen by this partition
+    // TODO: key_hash_index: KeyHashIndex,
```

Comment:
Not a blocker -- related to `Grouping` in `spread`, which we'll do as a follow-up